ipfx/qc_features.py

```diff
@@ -117,3 +111,31 @@ def get_r_from_peak_pulse_response(v, i, t):
         r.append(r_instance)
 
     return np.mean(r)
+
+
+def get_square_pulse_idx(v):
+    """
+    Get up and down indices of the square pulse(s).
+    Skipping the very first pulse (test pulse)
+
+    Parameters
+    ----------
+    v: float
+        pulse trace
+
+    Returns
+    -------
+    up_idx, down_idx: list, list
+        up, down indices
+    """
+    dv = np.diff(v)
+
+    up_idx = np.flatnonzero(dv > 0)[1:] # skip the very first pulse (test pulse)
+    down_idx = np.flatnonzero(dv < 0)[1:]
+
+    assert len(up_idx) == len(down_idx), "Truncated square pulse"
+
+    for up_ix, down_ix in zip(up_idx, down_idx):
+        assert up_ix < down_ix, "Negative pulse"
```

Comment:
Would it be useful to know if there are multiple instances of negative pulses? Or do you only care about whether the pulse trace has any negative pulses?

Reply:
I only care about the latter. If there is a negative pulse then there is a problem.